prql-php/src/Compiler.php

```diff
@@ -53,108 +92,94 @@ function compile(string $prql_query, ?Options $options = null): string
-                char* target;
-                bool signature_comment;
-            };
-
-            char* compile(const char *prql_query, struct options *opt);
-        ", __DIR__ . $library
-        );
-
-        $ffi_options = $libprql->new("struct options");
+        $ffi_options = $this->_libprql->new("struct Options");
         $ffi_options->format = $options->format;
         $ffi_options->signature_comment = $options->signature_comment;
 
         if (isset($options->target)) {
             $target_len = strlen($options->target);
-            $ffi_options->target = $ffi->new('char[$target_len]', 0);
-            FFI::memcpy($ffi_options->target, $options->target, $target_len);
-            FFI::free($ffi_options->target);
+            $ffi_options->target = \FFI::new("char[$target_len]", false);
+            \FFI::memcpy($ffi_options->target, $options->target, $target_len);
+            \FFI::free($ffi_options->target);
         }
 
         $out = str_pad("", 1024);
-        if ($libprql->compile($prql_query, \FFI::addr($$ffi_options)) !== 0) {
+        if ($this->_libprql->compile($prql_query, \FFI::addr($ffi_options), $out) !== 0) {
             throw new \InvalidArgumentException("Could not compile query.");
         }
 
+        unset($ffi_options);
+
         return trim($out);
     }
 
     /**
-     * Compile a PRQL string into a JSON string.
+     * Compile a PRQL string into PL.
      *
      * @param string $prql_query A PRQL query.
      *
-     * @return string JSON string.
+     * @return string Pipelined Language (PL) JSON string.
      * @throws \InvalidArgumentException If no query is given or the query cannot
      * be compiled.
      * @api
      */
-    function toJson(string $prql_query): string
+    function prqlToPL(string $prql_query): string
     {
         if (!$prql_query) {
             throw new \InvalidArgumentException("No query given.");
         }
 
-        $library = "/libprql_lib.so";
-
-        if (PHP_OS_FAMILY === "Windows") {
-            $library = "\libprql_lib.dll";
-        }
-
-        $libprql = \FFI::cdef(
-            "int to_json(char *prql_query, char *json_query);",
-            __DIR__ . $library
-        );
-
         $out = str_pad("", 1024);
-        if ($libprql->to_json($prql_query, $out) !== 0) {
+        if ($this->_libprql->prql_to_pl($prql_query, $out) !== 0) {
             throw new \InvalidArgumentException("Could not compile query.");
         }
 
         return trim($out);
     }
 
     /**
-     * Compile a PRQL string into a SQL string.
+     * Converts PL to RQ.
      *
-     * @param string $prql_query A PRQL query.
+     * @param string $pl_json PL in JSON format.
      *
-     * @return string SQL query.
+     * @return string RQ string.
      * @throws \InvalidArgumentException If no query is given or the query cannot
      * be compiled.
      * @api
      */
-    function toSql(string $prql_query): string
+    function pLToRQ(string $pl_json): string
```

Comment:
What's the convention in Java for upper case words?

In rust, the convention is to only ever have the first character of each word be upper case. So `trait Sql`, not `trait SQL`. That would be `plToRq` I guess?

(only matters on the external API, no need to change everything tbc)

Reply:
PHP uses camel case for method names and camel case is the recommendation by [PHP-FIG](https://www.php-fig.org/) and the configuration for [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) linter.

I don't know about acronyms, there aren't so many in PHP, but acronyms seem to be all uppercase in class names such as `SQLite3` and lowercase in method names such as [`Swoole\Async::dnsLookup`](https://www.php.net/manual/en/swoole-async.dnslookup) but that name starts with a lower case letter. PHP_CodeSniffer doesn't complain on method names that repeat capital letters.

I could leave `toSql` as is, or rename it to `toSQL`.
As for `pLToRQ` that one is trickier, but I should probably rename it to `plToRQ`maybe?

Reply:
Thanks for the links!

Yes, it should either be `plToRQ` or `plToRq`. My slight preference from Googling would be the latter — e.g. [here](https://stackoverflow.com/a/15526526/3064736), but totally fine with the former.

Reply:
Microsoft's recommendation for .NET is that acronyms with two letters (such as IO and UI) should have both letters uppercase, while acronyms longer two letters (such as TCP or HTTP) should be lowercase. So in the .NET world Microsoft would favor the former not the latter.

Reply:
I had thought that's for `foo.IO`, but not for `foo.IOTransport`, but you're completely right it's ambiguous, honestly either is fine, as long as we make the `pL` -> `pl`

Reply:
Done!